drivers: nxp_enet: Enable on k6x and RT11xx

decsny opened the pull request:
Enable NXP ENET driver for the freedom k64f board and RT11xx

Commit:
drivers: ksz8081: Some bug fixes & 25MHz RMII

- PHY can be set up as rmii but still use 25 MHz MDC, add DT property
  value for this case
- Fix KSZ8081 driver spamming phy status in debug level logging,
  and fix some other state/logging logic
- Fix PHY driver not rescheduling monitor work if first configuration
  fails, change code path to use goto for errors
- Handle case where some phys are not using the gpio pins in phy driver
  Make GPIO properties of ksz8081 phy optional since these hardware pins
  may be unused on some boards

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: nxp_enet: Do clock init from zephyr

Need to do the ENET module-level clock initialization from zephyr
instead of MCUX HAL, because now there are multiple zephyr drivers with
different init priorities that rely on the module being clocked. MDIO
must be initialized before the ENET MAC, and the MAC driver currently
calls ENET_Init from the HAL to initialize the clock, but MDIO needs the
module clock enabled first on some platforms. So replace the MAC init
with ENET_Up from the HAL, which doesn't include clock init, then do
clock init from a higher priority sys init based on the parent
compatible.

Also, add support for enet clock ungating with clock_control_on on ccm
driver do this with current platforms supported.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: clock_control: Support NXP_ENET

Support ENET peripheral clock in MCUX SIM and CCM_REV2 driver

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
soc: k6x: Support NXP ENET Driver

NXP ENET driver support from soc.c for k6x series

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

dleach02 commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -1040,4 +1040,29 @@ static const struct ethernet_api api_funcs = {
+	return 0;									\
+}											\
+											\
+	/* Init the module before any of the MAC, MDIO, or PTP clock */			\
```
Your macro continuation '\' went too far

dleach02 replied:
ah yes, I missed that

decsny commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -791,7 +791,7 @@ static int eth_nxp_enet_init(const struct device *dev)
 	enet_config.userData = (void *)dev;
 
-	ENET_Init(config->base,
+	ENET_Up(config->base,
```
I am not sure, but with this target, I noticed the debugger could not read the ENET registers during mdio initialization, and this change fixed that issue

hakehuang replied:
we do not call ENET_Init anymore, which would be dangous if SDK changes.

dleach02 replied:
Good point. We use the ENET_Init() to deal with platform specific stuff.

hakehuang commented on boards/arm/frdm_k64f/doc/index.rst:
```diff
@@ -378,3 +378,13 @@ of pyocd commands:
+enhanced capability, such as not hardcoding code for only one phy in the driver like eth_mcux.
+
+To build for this EVK with the new driver, include the experimental overlay to west build with
+the option `-DEXTRA_DTC_OVERLAY_FILE=nxp,enet-experimental.overlay`.
```
can we add this option as a new test case in one of the network testing sample.yaml file

decsny replied:
ok

Commit:
samples: zperf: Add NXP driver test case

Test the experimental NXP ENET driver

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

hakehuang commented:
test with zperf UDP, no issue, but I find the link detect does not work, the link status keeps on by default, no matter I plug in or plug our the ethernet cable. is this supported?
```
uart:~$ zperf udp upload 192.0.2.2 5001 10 1K 1000M
Remote port is 5001
Invalid IPv6 address 192.0.2.2
Connecting to 192.0.2.2
Duration:       10.00 s
Packet size:    1024 bytes
Rate:           1024000 kbps
Starting...
Rate:           1000.00 Mbps
Packet duration 7 us
-
Upload completed!
Statistics:             server  (client)
Duration:               10.00 s (10.00 s)
Num packets:            28752   (28752)
Num packets out order:  0
Num packets lost:       0
Jitter:                 45 us
Rate:                   22.46 Mbps      (22.46 Mbps)
```

decsny commented:
> @decsny test with zperf UDP, no issue, but I find the link detect does not work, the link status keeps on by default, no matter I plug in or plug our the ethernet cable. is this supported?

are you seeing this only on the k64?

Commit:
soc: rt11xx: support nxp_enet in soc

Support NXP ENET on RT11xx soc

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
boards: nxp: Support NXP ENET FRDM_K64F & RT11xx

Support NXP ENET on FRDM_K64F and rt11xx boards.
Document the experimental driver in all boards.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented:
> @decsny test with zperf UDP, no issue, but I find the link detect does not work, the link status keeps on by default, no matter I plug in or plug our the ethernet cable. is this supported?

okay, I see myself now too there is some issue with this on the kinetis board, which doesn't exist on the RT boards. thanks for reporting this

Commit:
drivers: mdio_nxp_enet: Fix typo in MDC frequency

MDC frequency had one extra 0 by mistake

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented:
@hakehuang I think I fixed the issues in the latest push, can you check

hakehuang commented:
> @hakehuang I think I fixed the issues in the latest push, can you check

@decsny , see below log, although the phy error is reported, but the link status is still on, although I did not plugin the ethernet cable, and after that no even triggered with cable in / out.

```
[00:00:04.553,000] <err> phy_mc_ksz8081: Error in autonegotiation
*** Booting Zephyr OS build zephyr-v3.5.0-2749-g903783a86f7d ***
[00:00:04.555,000] <inf> net_config: Initializing network
[00:00:04.555,000] <inf> net_config: IPv4 address: 192.0.2.1
[00:00:04.655,000] <inf> net_config: IPv6 address: 2001:db8::1
uart:~$ ifconfig
ifconfig: command not found
uart:~$ net iface

Interface eth0 (0x200012a4) (Ethernet) [1]
===================================
Link addr : 02:04:9F:51:82:96
MTU       : 1500
Flags     : AUTO_START,IPv4,IPv6
Device    : ethernet (0x25a70)
Ethernet capabilities supported:
        Virtual LAN
        10 Mbits
        100 Mbits
IPv6 unicast addresses (max 4):
        fe80::4:9fff:fe51:8296 autoconf preferred infinite
        2001:db8::1 manual preferred infinite
IPv6 multicast addresses (max 5):
        ff02::1
        ff02::1:ff51:8296
        ff02::1:ff00:1
IPv6 prefixes (max 2):
        <none>
IPv6 hop limit           : 64
IPv6 base reachable time : 30000
IPv6 reachable time      : 35882
IPv6 retransmit timer    : 0
IPv4 unicast addresses (max 1):
        192.0.2.1 manual preferred infinite
IPv4 multicast addresses (max 2):
        224.0.0.1
IPv4 gateway : 0.0.0.0
IPv4 netmask : 255.255.255.0
```

decsny commented:
> > @hakehuang I think I fixed the issues in the latest push, can you check
> 
> @decsny , see below log, although the phy error is reported, but the link status is still on, although I did not plugin the ethernet cable, and after that no even triggered with cable in / out.
> 
> ```
> [00:00:04.553,000] <err> phy_mc_ksz8081: Error in autonegotiation
> *** Booting Zephyr OS build zephyr-v3.5.0-2749-g903783a86f7d ***
> [00:00:04.555,000] <inf> net_config: Initializing network
> [00:00:04.555,000] <inf> net_config: IPv4 address: 192.0.2.1
> [00:00:04.655,000] <inf> net_config: IPv6 address: 2001:db8::1
> uart:~$ ifconfig
> ifconfig: command not found
> uart:~$ net iface
> 
> Interface eth0 (0x200012a4) (Ethernet) [1]
> ===================================
> Link addr : 02:04:9F:51:82:96
> MTU       : 1500
> Flags     : AUTO_START,IPv4,IPv6
> Device    : ethernet (0x25a70)
> Ethernet capabilities supported:
>         Virtual LAN
>         10 Mbits
>         100 Mbits
> IPv6 unicast addresses (max 4):
>         fe80::4:9fff:fe51:8296 autoconf preferred infinite
>         2001:db8::1 manual preferred infinite
> IPv6 multicast addresses (max 5):
>         ff02::1
>         ff02::1:ff51:8296
>         ff02::1:ff00:1
> IPv6 prefixes (max 2):
>         <none>
> IPv6 hop limit           : 64
> IPv6 base reachable time : 30000
> IPv6 reachable time      : 35882
> IPv6 retransmit timer    : 0
> IPv4 unicast addresses (max 1):
>         192.0.2.1 manual preferred infinite
> IPv4 multicast addresses (max 2):
>         224.0.0.1
> IPv4 gateway : 0.0.0.0
> IPv4 netmask : 255.255.255.0
> ```

can you clarify what part of this log is the problem ?

hakehuang commented:
> can you clarify what part of this log is the problem ?

@decsny , as I do not plugin the eth-cable, the link status should be none,  but it shows linked up. also, there used to be a LOG_INF to show the link status change, in eth_mcux, which is helpful usually for user to check the status. and it is gone.

Commit:
drivers: nxp_enet: Add link state logging message

Add info logging message to NXP ENET MAC driver about link state.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented:
@hakehuang There was a logging message about this from the PHY driver, but I also added one to the MAC driver, if you think it is useful.

hakehuang commented:
test with frdm_k64f and mimxrt1060_evk, zperf works fine

mmahadevan108 commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -1047,4 +1047,29 @@ static const struct ethernet_api api_funcs = {
+}											\
+											\
+	/* Init the module before any of the MAC, MDIO, or PTP clock */			\
+	SYS_INIT(nxp_enet_##n##_init, POST_KERNEL, 0);
```
Any reason this is tabbed in? Since it is not part of the function maybe the tab is not needed.

decsny replied:
it's part of the macro function

thedjnK commented on boards/arm/frdm_k64f/doc/index.rst:
```diff
@@ -378,3 +378,13 @@ of pyocd commands:
+enhanced capability, such as not hardcoding code for only one phy in the driver like eth_mcux.
+
+To build for this EVK with the new driver, include the experimental overlay to west build with
+the option `-DEXTRA_DTC_OVERLAY_FILE=nxp,enet-experimental.overlay`.
```
seems like this should be made into a snippet?
